Adding additional tests for pbsfs on multi scheds

latha-subramanian opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
Additional automated tests for verifying pbsfs on multi scheds 

#### Affected Platform(s)
* Linux

#### Cause / Analysis / Design
Missing tests on some pbsfs's error conditions and "fairshare tree trim" on multi scheds

#### Solution Description
Added automated tests for pbsfs's error conditions and "fairshare tree trim" on multi scheds in pbs_multi_sched.py

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

[pbs_multi_sched_tests.log](https://github.com/PBSPro/pbspro/files/1764694/pbs_multi_sched_tests.log)

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

anamikau commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -699,3 +699,128 @@ def test_cmp_fairshare(self):
+
+    def test_pbfs_server_restart(self):
+        """
+        Test pbsfs -I <sched_name> when server is restarted
```
update the description to "verify that server restart has no impact on fairshare usage"

anamikau commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -699,3 +699,128 @@ def test_cmp_fairshare(self):
+        self.server.manager(MGR_CMD_SET, SCHED, {'scheduling': 'True'},
+                            id='sc1')
+        self.scheds[sched_name].log_match(
+            'Leaving Scheduling Cycle', starttime=t_start, max_attempts=10)
```
remove max_attempts and use default (which is 30s)

anamikau commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -699,3 +699,128 @@ def test_cmp_fairshare(self):
+        t_end = int(time.time())
+        job_list = self.scheds[sched_name].log_match(
+            'Considering job to run', starttime=t_start,
+            allmatch=True, max_attempts=10, endtime=t_end)
```
again remove max_attempts and use default

anamikau commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -699,3 +699,128 @@ def test_cmp_fairshare(self):
+        self.server.manager(MGR_CMD_SET, SCHED, {'scheduling': 'True'},
+                            id='sc1')
+        self.scheds[sched_name].log_match(
+            'Leaving Scheduling Cycle', starttime=t_start, max_attempts=10)
```
remove max_attempts

anamikau commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -699,3 +699,128 @@ def test_cmp_fairshare(self):
+        t_end = int(time.time())
+        job_list = self.scheds[sched_name].log_match(
+            'Considering job to run', starttime=t_start,
+            allmatch=True, max_attempts=10, endtime=t_end)
```
remove max_attempts

bhroam commented:
Codacy is complaining about a few unused variables.  Please remove those.

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -699,3 +699,128 @@ def test_cmp_fairshare(self):
+        Test pbsfs's trim fairshare tree functionality for multisched
+        """
+        self.common_setup()
+        sched_name = 'sc1'
```
Does setup for test needs all scheduers ? else you can just setup_sc1 and queues and nodes as common_setup is heavy . 
Is there any intension of using sched_name variable than just using 'sc1' at all places ?

latha-subramanian replied:
@visheshh . No. There wasn't any specific reason. I have reverted to 'sc1' as commonly done in other tests of the file.  Also, creating 'sc1' specific sched, queue and node. Got rid of common_setup(). Thanks!

latha-subramanian commented:
@bhroam, @anamikau and @visheshh , I have addressed all of your review comments. Please have a look.  Thanks!

[pbs_multi_sched.log](https://github.com/PBSPro/pbspro/files/1768440/pbs_multi_sched.log)

bhroam commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -754,18 +742,30 @@ def test_pbfs_server_restart(self):
+        self.server.manager(MGR_CMD_CREATE, QUEUE, a, id='wq1')
+
+        p1 = {'partition': 'P1'}
+        self.server.manager(MGR_CMD_SET, QUEUE, p1, id='wq1', expect=True)
```
Combine the above two manager calls.  Put partition in the initial set of attributes.

bhroam commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -754,18 +742,30 @@ def test_pbfs_server_restart(self):
+        p1 = {'partition': 'P1'}
+        self.server.manager(MGR_CMD_SET, QUEUE, p1, id='wq1', expect=True)
+        a = {'resources_available.ncpus': 2}
+        self.server.create_vnodes('vnode', a, 1, self.mom)
```
Instead of creating one vnode, you could just use the mom directly.  Use a manager on self.mom.shortname

latha-subramanian commented:
@bhroam , I have addressed your comments. Thanks!

[test_pbsfs_revert_to_defaults.log](https://github.com/PBSPro/pbspro/files/1776602/test_pbsfs_revert_to_defaults.log)

bhroam commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -751,14 +751,12 @@ def test_pbsfs_revert_to_defaults(self):
-        self.server.create_vnodes('vnode', a, 1, self.mom)
-        self.server.manager(MGR_CMD_SET, NODE, p1, id='vnode[0]', expect=True)
+        self.server.manager(MGR_CMD_SET, NODE, {'partition': 'P1'},
+                            id=self.mom.shortname, expect=True)
```
Shouldn't this be partition and resources_available.ncpus?

latha-subramanian replied:
Oops! I missed it.  Changed it now. Please review.
[test_pbsfs_revert_to_defaults.log](https://github.com/PBSPro/pbspro/files/1776656/test_pbsfs_revert_to_defaults.log)

bhroam commented:
Please squash and rebase for the merge.

latha-subramanian commented:
@bhroam,  I have squashed and rebased.  Please take it forward.

bhroam commented:
I merged in the PR.
